### PR TITLE
feat: filter by all version states

### PR DESCRIPTION
**Pull request by Bernardvdv**
Adding filter by all states to admin

### CODE REVIEW

**Review by Aiky30 — CHANGES_REQUESTED** [review]
Looks good. It will need a test for each filter state i.e. All, Draft, Published, Archived, Unpublished. You have published and draft currently.

File: tests/test_views.py [hunk 1]
```diff
@@ -158,44 +160,145 @@ def test_view_poll_plugin_attached_to_page_should_return_related_page(self):
...
+        # add poll plugin to page
+        add_plugin(placeholder1, "PollPlugin", "en", poll=poll1, template=0)
+        add_plugin(placeholder1, "PollPlugin", "en", poll=poll1, template=0)
+
+        with self.login_user_context(self.superuser):
+            url = self.get_view_url(
+                    content_type_id=ContentType.objects.get_for_model(poll1).pk,
+                    object_id=poll1.id,
+                ) + "?state=published"
+
```
Comment (adam-murray) [inline tests/test_views.py:232]: This doesn't look like it follows AAA, could you remove the space here?

**Review by adam-murray — CHANGES_REQUESTED** [review]
Looks good, but should follow AAA: https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/

File: djangocms_references/helpers.py [hunk 2]
```diff
@@ -257,7 +243,14 @@ def apply_additional_modifiers(queryset):
     return queryset
 
 
-def get_all_reference_objects(content, draft_and_published=False):
+def apply_filters(queryset, model, state_selected):
+    if get_versionable_for_content(model):
```
Comment (Aiky30) [inline djangocms_references/helpers.py:247]:
```suggestion
    if get_versionable_for_content(queryset.model):
```
The model can be different per queryset, not all items will be versionable, by using the user supplied model we are assuming every queryset is of the same model that we are looking for references for. 

For example, I'm looking for all references of a Page, The model is then Page. The Page could be found inside Alias, Navigation, some other none versioned model etc. Each queryset the model would then be:  Alias, Navigation, some other model. Not Page which is what this is currently doing.

File: djangocms_references/helpers.py [hunk 3]
```diff
@@ -257,7 +243,14 @@ def apply_additional_modifiers(queryset):
...
+        queryset = queryset.filter(versions__state__in=([state_selected]))
+        return queryset
+    return queryset
+
+
+def get_all_reference_objects(content, state_selected=None, model=None):
```
Comment (Aiky30) [inline djangocms_references/helpers.py:253]: The params have changed but the documentation within the function has been neglected. This package has been very well documented, it's how code should be documented and something that we should be doing with any new code.

File: CHANGELOG.rst [hunk 4]
```diff
@@ -4,6 +4,7 @@ Changelog
 
 Unreleased
 ==========
+* Added filter by latest state
```
Comment (Aiky30) [inline CHANGELOG.rst:7]:
```suggestion
* feat: Added filter by latest state
```

Prefix using conventional commits style: https://www.conventionalcommits.org/en/v1.0.0-beta.2/

File: tests/test_views.py [hunk 5]
```diff
@@ -158,44 +160,145 @@ def test_view_poll_plugin_attached_to_page_should_return_related_page(self):
...
+                    content_type_id=ContentType.objects.get_for_model(poll1).pk,
+                    object_id=poll1.id,
+                ) + "?state=unpublised"
+
+            response = self.client.get(url)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context["querysets"][0].count(), 1)
+        self.assertNotIn(page_content1, response.context["querysets"][0])
+        self.assertIn(page_content2, response.context["querysets"][0])
```
Comment (Aiky30) [inline tests/test_views.py:312]: Missing a test for the "all" filter

**Review by Aiky30 — CHANGES_REQUESTED** [review]
A few changes required, hopefully fairly trivial.

File: tests/test_views.py (hunk 1 above)
Comment (Bernardvdv) [reply tests/test_views.py:235]: @adam-murray  Thanks, updated the test cases to follow AAA

File: tests/test_views.py [hunk 6]
```diff
@@ -158,44 +165,194 @@ def test_view_poll_plugin_attached_to_page_should_return_related_page(self):
...
+            object_id=page_content1.id,
         )
-        add_plugin(placeholder1, "PollPlugin", "en", poll=poll, template=0)
 
-        version2 = archived.copy(archived.created_by)
-        version2.publish(archived.created_by)
-        version3 = version2.copy(version2.created_by)
+        poll1 = PollFactory()
+        # add poll plugin to page
+        add_plugin(placeholder1, "PollPlugin", "en", poll=poll1, template=0)
```
Comment (Aiky30) [inline tests/test_views.py:196]: There's some inconsistencies in the tests. This test will need an additional plugin as per the others.

File: tests/test_views.py [hunk 7]
```diff
@@ -158,44 +165,194 @@ def test_view_poll_plugin_attached_to_page_should_return_related_page(self):
...
+        page_content2 = version2.content
+
+        placeholder1 = PlaceholderFactory(
+            content_type=ContentType.objects.get_for_model(page_content2),
+            object_id=page_content2.id,
+        )
+
+        poll1 = PollFactory()
+        # add poll plugin to page
+        add_plugin(placeholder1, "PollPlugin", "en", poll=poll1, template=0)
```
Comment (Aiky30) [inline tests/test_views.py:267]: There's some inconsistencies in the tests. This test will need an additional plugin as per the others.

File: tests/test_views.py [hunk 8]
```diff
@@ -158,44 +165,194 @@ def test_view_poll_plugin_attached_to_page_should_return_related_page(self):
...
```
Comment (Aiky30) [inline tests/test_views.py:302]: There's some inconsistencies in the tests. This test will need an additional plugin as per the others.

**Review by Aiky30 — APPROVED** [review]

**Pull request merged**